internal/server/data/grant.go

```diff
@@ -106,8 +104,91 @@ func GetGrant(tx ReadTxn, opts GetGrantOptions) (*models.Grant, error) {
 	return (*models.Grant)(table), nil
 }
 
-func ListGrants(db GormTxn, p *Pagination, selectors ...SelectorFunc) ([]models.Grant, error) {
-	return list[models.Grant](db, p, selectors...)
+type ListGrantsOptions struct {
+	BySubject   uid.PolymorphicID
+	ByPrivilege string
+	ByResource  string
+
+	// IncludeInheritedFromGroups instructs ListGrants to include grants from
+	// groups where the user is a member. This option can only be used when
+	// BySubject is a non-zero userID.
+	IncludeInheritedFromGroups bool
+
+	// IncludeConnectorGrant instructs ListGrants to exclude grants where the
+	// CreatedBy is models.CreatedBySystem.
+	IncludeConnectorGrant bool
+
+	Pagination *Pagination
+}
+
+func ListGrants(tx ReadTxn, opts ListGrantsOptions) ([]models.Grant, error) {
+	table := grantsTable{}
+	query := querybuilder.New("SELECT")
+	query.B(columnsForSelect(table))
+	if opts.Pagination != nil {
+		query.B(", count(*) OVER()")
+	}
+	query.B("FROM grants")
+	query.B("WHERE deleted_at is null")
+	query.B("AND organization_id = ?", tx.OrganizationID())
+
+	if opts.BySubject != "" {
+		if !opts.IncludeInheritedFromGroups {
+			query.B("AND subject = ?", opts.BySubject)
+		} else {
+			subjects := []string{opts.BySubject.String()}
+
+			userID, err := opts.BySubject.ID()
+			if err != nil || !opts.BySubject.IsIdentity() {
+				return nil, fmt.Errorf("IncludeInheritedFromGroups requires a userId subject")
+			}
+			// FIXME: store userID and groupID as a field on the grants table so
+			// that we can replace this with a sub-select or join.
+			groupIDs, err := groupIDsForUser(tx, userID)
+			if err != nil {
+				return nil, err
+			}
+			for _, id := range groupIDs {
+				subjects = append(subjects, uid.NewGroupPolymorphicID(id).String())
+			}
+			query.B(`AND subject IN (?)`, subjects)
+		}
+	}
+	if opts.ByPrivilege != "" {
+		query.B("AND privilege = ?", opts.ByPrivilege)
+	}
+	if opts.ByResource != "" {
+		query.B("AND resource = ?", opts.ByResource)
+	}
+	if !opts.IncludeConnectorGrant {
+		query.B("AND NOT (privilege = 'connector' AND resource = 'infra')")
+	}
```

Comment:
In some of the intermediate commits I tried to change this to `IncludeCreatedBySystem`, and use `CreatedBy <> 1`. Unfortunately that did not work out because we create grants from the config using the same `CreatedBy` ID. It would be nice if we could exclude system grants in a more general way.